[Bug]: Context window exceeded error not triggering retry logic during episodic summarization (original error message missing in ExternalServiceAPIError)

### Describe the bug

When the LLM returns a context length exceeded error, MemMachine’s retry logic fails to activate. The root cause is that ExternalServiceAPIError is raised without including the original error message from the LLM. As a result, _is_exceed_context_window_error() cannot detect keywords like “context length” or “token limit”, skipping the split-and-retry mechanism.
This leads to silent failure with “Summarization failed due to unexpected error”, risking data loss.

```
2026-03-18 19:43:14,833 [ERROR] memmachine_server.common.language_model.openai_chat_completions_language_model - [call uuid: e3758a9d-e2ef-481c-b14c-f381949b03e1] Giving up generating response after failed attempt 1 due to non-retryable BadRequestError
Traceback (most recent call last):
File "/path/to/mm/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 219, in _generate_response
response = await self._client.chat.completions.create(**args)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 2676, in create
return await self._post(
^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1669, in request
raise self._make_status_error_from_response(err.response) from None
openai.BadRequestError: Error code: 400 - {'error': {'message': "You passed 202753 input tokens and requested 0 output tokens. However, the model's context length is only 202752 tokens, resulting in a maximum input length of 202752 tokens. Please reduce the length of the input prompt. (parameter=input_tokens, value=202753)", 'type': 'BadRequestError', 'param': 'input_tokens', 'code': 400}}
2026-03-18 19:43:14,836 [ERROR] memmachine_server.episodic_memory.short_term_memory.short_term_memory - Summarization failed due to unexpected error
Traceback (most recent call last):
File "/path/to/mm/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 219, in _generate_response
response = await self._client.chat.completions.create(**args)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/resources/chat/completions/completions.py", line 2676, in create
return await self._post(
^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1884, in post
return await self.request(cast_to, opts, stream=stream, stream_cls=stream_cls)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/.venv/lib/python3.12/site-packages/openai/_base_client.py", line 1669, in request
raise self._make_status_error_from_response(err.response) from None
openai.BadRequestError: Error code: 400 - {'error': {'message': "You passed 202753 input tokens and requested 0 output tokens. However, the model's context length is only 202752 tokens, resulting in a maximum input length of 202752 tokens. Please reduce the length of the input prompt. (parameter=input_tokens, value=202753)", 'type': 'BadRequestError', 'param': 'input_tokens', 'code': 400}}

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/path/to/mm/packages/server/src/memmachine_server/episodic_memory/short_term_memory/short_term_memory.py", line 606, in _create_summary
result = await self._model.generate_response(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 160, in generate_response
output, function_calls_arguments, _, _ = await self._generate_response(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/path/to/mm/packages/server/src/memmachine_server/common/language_model/openai_chat_completions_language_model.py", line 259, in _generate_response
raise ExternalServiceAPIError(error_message) from e
memmachine_server.common.data_types.ExternalServiceAPIError: [call uuid: e3758a9d-e2ef-481c-b14c-f381949b03e1] Giving up generating response after failed attempt 1 due to non-retryable BadRequestError
```
 
> Note: We are not using OpenAI’s cloud LLM. We are using an on-premise LLM that exposes an OpenAI-compatible API endpoint (vLLM). The error structure and behavior are identical to OpenAI’s, but the service is self-hosted. 

### Steps to reproduce

1. Start MemMachine server.
2. Use REST client to rapidly inject [BEAM benchmark](https://github.com/mohammadtavakoli78/BEAM) conversations (20x ~100K token chats, 35x ~500K token chats, 35x ~1M token chats) via add(...) with both episodic & semantic memory enabled in a loop.
3. When LLM context window is exceeded (e.g., 202753 tokens vs. 202752 limit), observe:
- LLM returns BadRequestError with message: “You passed 202753 input tokens... context length is 202752”
- MemMachine logs ExternalServiceAPIError without original message
- No “Batch failed. Splitting X episodes into halves.” log
- Final error: “Summarization failed due to unexpected error”

### Expected behavior

- On context window exceedance, _is_exceed_context_window_error() should detect “context length” in the error message.
- Split-and-retry logic should trigger: log “Batch failed. Splitting X episodes into halves.” and retry with halved input.
- Original LLM error message must be preserved in ExternalServiceAPIError for accurate detection.

### Environment

- OS: Ubuntu 22.04.5 LTS 
- MemMachine Version: v0.3.3.dev1+g75e871eb6
- Development language version: Python 3.12
- LLM Backend: On-premise LLM with OpenAI API compatibility (vLLM)

### Additional context

I fixed this myself and it’s working well in our environment. If you think it’s safe, feel free to merge it directly — it’s just one line.
(Not sure if this bug also exists when using the official OpenAI cloud service — but since we’re on-prem, this fix is critical for us.)

Root Cause:
- In `openai_chat_completions_language_model.py`, ExternalServiceAPIError is raised with a generic message:

```python
original_error_msg = str(e)
error_message = (
    f"[call uuid: {generate_response_call_uuid}] "
    "Giving up generating response "
    f"after failed attempt {attempt} "
    f"due to non-retryable {type(e).__name__}"  # before
)
```

- The original LLM error (e.g., “You passed 202753 input tokens... context length is 202752”) is omitted.
→ `_is_exceed_context_window_error()` receives no keywords → returns False → retry logic skipped.

Fix:
- Include original error message in ExternalServiceAPIError:

```python
original_error_msg = str(e)
error_message = (
    f"[call uuid: {generate_response_call_uuid}] "
    "Giving up generating response "
    f"after failed attempt {attempt} "
    f"due to non-retryable {type(e).__name__}: {original_error_msg}"   # after
)
```

Verification:
- After applying the fix, context window errors from the LLM now correctly trigger split-and-retry logic. No summary data loss observed.

--- a/packages/server/server_tests/memmachine_server/episodic_memory/short_term_memory/test_short_term_memory.py
+++ b/packages/server/server_tests/memmachine_server/episodic_memory/short_term_memory/test_short_term_memory.py
@@ -348,6 +348,78 @@ class TestSessionMemoryPublicAPI:
         assert mock_model.call_count == 4
 
     @pytest.mark.asyncio
+    async def test_summary_exceed_context_window_single_episode(
+        self, mock_data_manager
+    ):
+        class ChunkingLanguageModel(LanguageModel):
+            def __init__(self) -> None:
+                self.call_count = 0
+
+            async def generate_response(
+                self,
+                system_prompt: str | None = None,
+                user_prompt: str | None = None,
+                tools: list[dict[str, Any]] | None = None,
+                tool_choice: str | dict[str, str] | None = None,
+                max_attempts: int = 1,
+            ) -> tuple[str, Any]:
+                prompt = user_prompt or ""
+                if len(prompt) > 10000:
+                    raise ValueError("User prompt exceeds context window")
+                self.call_count += 1
+                match = re.search(r"summary:(\d+)(?: \d+)?$", prompt)
+                prev_summary = int(match.group(1)) if match else 0
+                messages = re.findall(r'"([^"]*)"', prompt)
+                return (
+                    f"summary:{prev_summary + sum(len(message) for message in messages)}",
+                    "",
+                )
+
+            async def generate_response_with_token_usage(
+                self,
+                system_prompt: str | None = None,
+                user_prompt: str | None = None,
+                tools: list[dict[str, Any]] | None = None,
+                tool_choice: str | dict[str, str] | None = None,
+                max_attempts: int = 1,
+            ) -> tuple[str, Any, int, int]:
+                response, tool_output = await self.generate_response(
+                    system_prompt=system_prompt,
+                    user_prompt=user_prompt,
+                    tools=tools,
+                    tool_choice=tool_choice,
+                    max_attempts=max_attempts,
+                )
+                return response, tool_output, 0, 0
+
+            async def generate_parsed_response(
+                self,
+                output_format: type[T],
+                system_prompt: str | None = None,
+                user_prompt: str | None = None,
+                max_attempts: int = 1,
+            ) -> T:
+                return cast(T, "summary")
+
+        model = ChunkingLanguageModel()
+        params = ShortTermMemoryConsolidator.Params(
+            summary_user_prompt="User prompt: {episodes} {summary} {max_length}",
+            summary_system_prompt="System Prompt",
+            max_summary_length_words=100,
+            session_key="test_session",
+            model=model,
+            data_manager=mock_data_manager,
+        )
+        consolidator = ShortTermMemoryConsolidator(params)
+
+        oversized_content = "a" * 12000
+        episode = create_test_episode(content=oversized_content)
+        summary = await consolidator._create_summary("", [episode])
+
+        assert summary == "summary:12000"
+        assert model.call_count > 1
+
+    @pytest.mark.asyncio
     async def test_summary_catch_up(self, mock_model, mock_data_manager):
         params = ShortTermMemoryConsolidator.Params(
             summary_user_prompt="User prompt: {episodes} {summary} {max_length}",

--- a/packages/server/src/memmachine_server/episodic_memory/short_term_memory/short_term_memory.py
+++ b/packages/server/src/memmachine_server/episodic_memory/short_term_memory/short_term_memory.py
@@ -580,11 +580,31 @@ class ShortTermMemoryConsolidator:
         ]
         return any(keyword in error_msg for keyword in keywords)
 
+    @staticmethod
+    def _split_episode(episode: Episode) -> tuple[Episode, Episode] | None:
+        """
+        Split a single oversized episode into two synthetic chunks.
+
+        The chunks keep the original episode metadata so they can be summarized
+        sequentially as if they were adjacent messages from the same source.
+        """
+        if len(episode.content) <= 1:
+            return None
+
+        mid = len(episode.content) // 2
+        if mid == 0 or mid >= len(episode.content):
+            return None
+
+        first_chunk = episode.model_copy(update={"content": episode.content[:mid]})
+        second_chunk = episode.model_copy(update={"content": episode.content[mid:]})
+        return first_chunk, second_chunk
+
     async def _create_summary(self, summary: str, episodes: list[Episode]) -> str:
         """
         Generate a summary recursively.
 
-        splitting the batch if it exceeds the context window or encounters an error.
+        Split the work into smaller pieces when the prompt exceeds the model's
+        context window so all episode content is still consumed.
         """
         # Base Case: Nothing to process
         if not episodes:
@@ -616,26 +636,30 @@ class ShortTermMemoryConsolidator:
                 )
         except (ExternalServiceAPIError, ValueError, RuntimeError) as e:
             if self._is_exceed_context_window_error(e):
-                # If a single episode fails, drop it and return the current summary
+                batches: list[list[Episode]]
                 if len(episodes) == 1:
-                    logger.exception("Dropping failed episode %s", episodes[0].uid)
-                    return summary
+                    split_episode = self._split_episode(episodes[0])
+                    if split_episode is None:
+                        logger.exception("Dropping failed episode %s", episodes[0].uid)
+                        return summary
+                    logger.warning(
+                        "Episode %s exceeded context window. Splitting content into "
+                        "smaller chunks.",
+                        episodes[0].uid,
+                    )
+                    batches = [[split_episode[0]], [split_episode[1]]]
+                else:
+                    mid = len(episodes) // 2
+                    logger.warning(
+                        "Batch failed. Splitting %d episodes into halves.",
+                        len(episodes),
+                    )
+                    batches = [episodes[:mid], episodes[mid:]]
 
-                # Otherwise, split and recurse
-                mid = len(episodes) // 2
-                logger.warning(
-                    "Batch failed. Splitting %d episodes into halves.", len(episodes)
-                )
-
-                # 1. Summarize the first half
-                summary_after_first_half = await self._create_summary(
-                    summary, episodes[:mid]
-                )
-
-                # 2. Use that result as the 'base' to summarize the second half
-                return await self._create_summary(
-                    summary_after_first_half, episodes[mid:]
-                )
+                current_summary = summary
+                for batch in batches:
+                    current_summary = await self._create_summary(current_summary, batch)
+                return current_summary
             # For other errors, log and ignore
             logger.exception("Summarization failed due to unexpected error")
         else:

